hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotCache.java

```diff
@@ -0,0 +1,323 @@
+    release(key);
+  }
+
+  /**
+   * Callback method used to enqueue or dequeue ReferenceCounted from
+   * pendingEvictionList.
+   * @param referenceCounted ReferenceCounted object
+   */
+  public void callback(ReferenceCounted referenceCounted) {
+    if (referenceCounted.getTotalRefCount() == 0L) {
+      // Reference count reaches zero, add to pendingEvictionList
+      Preconditions.checkState(!pendingEvictionList.contains(referenceCounted),
+          "SnapshotCache is inconsistent. Entry should not be in the "
+              + "pendingEvictionList when ref count just reached zero.");
+      pendingEvictionList.add(referenceCounted);
+    } else if (referenceCounted.getTotalRefCount() == 1L) {
+      pendingEvictionList.remove(referenceCounted);
+    }
+  }
+
+  /**
+   * If cache size exceeds soft limit, attempt to clean up and close the
+   * instances that has zero reference count.
+   * TODO: [SNAPSHOT] Add new ozone debug CLI command to trigger this directly.
+   */
+  private void cleanup() {
+    long numEntriesToEvict = (long) dbMap.size() - cacheSizeLimit;
+    while (pendingEvictionList.size() > 0 && numEntriesToEvict > 0L) {
+      // Get the first instance in the clean up list
+      ReferenceCounted<IOmMetadataReader> rcOmSnapshot =
+          pendingEvictionList.iterator().next();
+      OmSnapshot omSnapshot = (OmSnapshot) rcOmSnapshot.get();
+      LOG.debug("Evicting OmSnapshot instance {} with table key {}",
+          rcOmSnapshot, omSnapshot.getSnapshotTableKey());
+      // Sanity check
+      Preconditions.checkState(rcOmSnapshot.getTotalRefCount() == 0L,
+          "Illegal state: OmSnapshot reference count non-zero ("
+              + rcOmSnapshot.getTotalRefCount() + ") but shows up in the "
+              + "clean up list");
+
+      final String key = omSnapshot.getSnapshotTableKey();
+      final ReferenceCounted<IOmMetadataReader> result = dbMap.remove(key);
+      // Sanity check
+      Preconditions.checkState(rcOmSnapshot == result,
+          "Cache map entry removal failure. The cache is in an inconsistent "
+              + "state. Expected OmSnapshot instance: " + rcOmSnapshot
+              + ", actual: " + result);
+
+      pendingEvictionList.remove(result);
+
+      // Close the instance, which also closes its DB handle.
+      try {
+        ((OmSnapshot) rcOmSnapshot.get()).close();
+      } catch (IOException ex) {
+        throw new IllegalStateException("Error while closing snapshot DB", ex);
+      }
+
+      --numEntriesToEvict;
+    }
+
+    // Print warning message if actual cache size is exceeding the soft limit
+    // even after the cleanup procedure above.
+    if ((long) dbMap.size() > cacheSizeLimit) {
+      LOG.warn("Current snapshot cache size ({}) is exceeding configured "
+          + "soft-limit ({}) after possible evictions.",
+          dbMap.size(), cacheSizeLimit);
+
+      Preconditions.checkState(pendingEvictionList.size() == 0);
```

Comment:
I'm skeptical about it and would prefer to add it in if block because there is no synchronization between this and `pendingEvictionList.add(referenceCounted);` at line # 249.

Reply:
`cleanup()` calls are now wrapped inside `synchronized (pendingEvictionList)` by callers.